File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/domain/policies/RefundGiftCardPolicyTest.java
package it.giannibombelli.wsc2026.giftcard.domain.policies;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.application.commands.RefundGiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static it.giannibombelli.wsc2026.testsupport.AggregateFactory.createBooking;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class RefundGiftCardPolicyTest {

    private final RefundGiftCardPolicy policy = new RefundGiftCardPolicy();

    @Test
    void evaluate_onBookingRejected_returnsRefundGiftCard() {
        GiftCardId giftCardId = EntityId.generate(GiftCardId::new);
        Money amount = new Money(new BigDecimal("50.00"));
        Booking booking = createBooking();
        BookingResultEvents.BookingRejected event = booking.reject(giftCardId, amount);

        RefundGiftCard result = policy.evaluate(event);

        assertThat(result).isNotNull();
        assertThat(result.aggregateId()).isEqualTo(giftCardId);
        assertThat(result.amount()).isEqualTo(amount);
    }

    @Test
    void evaluate_onBookingConfirmed_returnsNull() {
        GiftCardId giftCardId = EntityId.generate(GiftCardId::new);
        Money amount = new Money(new BigDecimal("100.00"));
        Booking booking = createBooking();
        BookingResultEvents.BookingConfirmed event = (BookingResultEvents.BookingConfirmed) booking.confirm(giftCardId, amount);

        RefundGiftCard result = policy.evaluate(event);

        assertThat(result).isNull();
    }

    @Test
    void evaluate_onBookingRefused_returnsNull() {
        GiftCardId giftCardId = EntityId.generate(GiftCardId::new);
        Money amount = new Money(new BigDecimal("75.50"));
        Booking booking = createBooking();
        BookingResultEvents.BookingRefused event = new BookingResultEvents.BookingRefused(booking.id(), giftCardId, amount);

        RefundGiftCard result = policy.evaluate(event);

        assertThat(result).isNull();
    }

    @Test
    void evaluate_onNullEvent_throwsException() {
        assertThatThrownBy(() -> policy.evaluate(null))
            .isInstanceOf(IllegalArgumentException.class);
    }
}

File: typescript/tests/giftcard/application/policies/refundGiftCardPolicy.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import {
  bookingConfirmed,
  bookingRejected,
  bookingRefused,
} from '@/booking/domain/events/bookingResultEvents.js';
import { RefundGiftCardPolicy } from '@/giftcard/application/policies/refundGiftCardPolicy.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { GiftCardAggregateFactory } from '../../../testsupport/giftcard/aggregateFactory.js';

describe('RefundGiftCardPolicy', () => {
  const policy = new RefundGiftCardPolicy();

  it('on BookingRejected returns RefundGiftCard', () => {
    const giftCardId = generateId((value) => new GiftCardId(value));
    const amount = new Money(50);
    const bookingId = GiftCardAggregateFactory.createBooking();
    const event = bookingRejected(bookingId, giftCardId, amount);

    const result = policy.evaluate(event);

    expect(result).not.toBeNull();
    expect(result!.aggregateId).toEqual(giftCardId);
    expect(result!.amount).toEqual(amount);
  });

  it('on BookingConfirmed returns null', () => {
    const giftCardId = generateId((value) => new GiftCardId(value));
    const amount = new Money(100);
    const bookingId = GiftCardAggregateFactory.createBooking();
    const event = bookingConfirmed(bookingId, giftCardId, amount);

    const result = policy.evaluate(event);

    expect(result).toBeNull();
  });

  it('on BookingRefused returns null', () => {
    const giftCardId = generateId((value) => new GiftCardId(value));
    const amount = new Money(75.5);
    const bookingId = GiftCardAggregateFactory.createBooking();
    const event = bookingRefused(bookingId, giftCardId, amount);

    const result = policy.evaluate(event);

    expect(result).toBeNull();
  });

  it('on null event throws', () => {
    expect(() => policy.evaluate(null as unknown as Parameters<typeof policy.evaluate>[0])).toThrow();
  });
});
